acceptance-tests/src/test/java/uk/nhs/digital/ps/test/acceptance/pages/LoginPage.java

```diff
@@ -54,12 +121,19 @@ public String findLoginErrorMessage() {
         return helper.findChildElement(hippoLoginFeedbackPanel, By.tagName("span")).getText();
     }
 
+    private void waitForLoginToComplete(final WebDriver webDriver) {
+        findLogoutButton(webDriver);
```

Comment:
This call isn't blocking, the logout button will be evaluated to null instantly as you are calling findOptionalElement

Reply:
Actually, in the context it's being used in, it will block and this has been verified to be happening. In  performance tests (the only context where this method is being used) a custom instance of WebDriver is being used, one with so called '[implicit wait](https://www.seleniumhq.org/docs/04_webdriver_advanced.jsp)' configured (as produced by `WebDriverProvider#newWebDriver`).

We don't use the implicit waits in 'normal' acceptance tests because we need more fine-grained control, but in performance tests their use allowed to reduce boilerplate code by delegating polling for conditions to the WebDriver itlsef.

Reply:
Ah I see, my worry is that someone will use this method in normal acceptance tests thinking it will block and it won't under normal conditions. Is there any way to make this clearer do you think?

Reply:
That is a good point and a valid concern though, on closer look, it seems that the non-blocking behaviour _is_ the one displayed by the original version as both the original version and my overload, rely on the same `PageHelper.findOptionalElement > PageHelper.findFirstOrNull > WebDriver.findElements` `PageHelper` methods.

It therefore appears that whichever version someone ends up using will continue to work as it used to, as my changes didn't really alter anything in the way `findLogoutButton` always worked.

Having said that, your observation prompted me to think that perhaps it would be better to change either the `waitForLoginToComplete` or the overloads of `findLogoutButton` to actually be blocking. Funny thing: git history shows that this method used to be blocking but was then changed to start using the non-blocking `PageHelper.findOptionalElement`. In the event that this was incidental, I'll give making it properly blocking and see what impact this has on the tests - both the acceptance and performance ones.

Reply:
Eventually decided to just make the `waitForLoginToComplete` more explicit in its blocking (and make it blocking even should someone want to use it with non-waiting, default version of WebDriver).

The `findLogoutButton` method needs to remain non-blocking in normal acceptance tests due to its use in `isLoggedIn` method; switching the former to use `WebDriver.findElement` would make it throw exception on element not found, while `WebDriver.findElements` only returns `null` in such a case without throwing anything which is what we want here.

Reply:
👍  nice one